### PR TITLE
Add support for aten::remainder.Tensor_out for MPS backend

### LINKED ISSUE #86806: [MPS] Add support for aten::remainder.Tensor_out for MPS backend
### 🐛 Describe the bug

First time contributors are welcome! 🙂
Add support for [aten::remainder.Tensor_out](https://pytorch.org/docs/stable/generated/torch.remainder.html) for MPS backend. Generic support for adding operations to MPS backend is captured here: https://github.com/pytorch/pytorch/wiki/MPS-Backend#adding-op-for-mps-backend

### Versions

N/A

cc @kulinseth @albanD @malfet @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by yash-dani**
Fixes #86806

### CODE REVIEW

**Comment by gpomeranz** [comment]
Hi @yash-dani , thanks so much for working on this! I seem to be having this exact issue when running pytorch on my M1, but the build you suggested is failing, same as above. Any chacne you could have a second to look at it?

Thank you so much for doing this :)

Kind regards,
Gideon

File: aten/src/ATen/native/mps/operations/UnaryOps.mm [hunk 1]
```diff
@@ -231,5 +231,24 @@ void unary_op(const Tensor& self, const Tensor& output, std::string op_name, Una
                 });
 }
 
+
+TORCH_IMPL_FUNC(remainder_out_mps) (const Tensor& self, const Tensor& output) {
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/UnaryOps.mm:235]: This function definition is not correct. Infact it should be in BinaryOps.mm. 
Please take a look at the formula to implement:
https://pytorch.org/docs/stable/generated/torch.remainder.html

**Review by kulinseth — CHANGES_REQUESTED** [review]

**Comment by kulinseth** [comment]
I have attached the patch , which should work.
[patch.txt](https://github.com/pytorch/pytorch/files/10038921/patch.txt)

**Comment by kulinseth** [comment]
@yash-dani , can you rebase and apply the patch.

**Comment by gpomeranz** [comment]
I've also send a pull request to @yash-dani where I rebased his repo and added the patch.

**Comment by yash-dani** [comment]
I can rebase and merge!

**Comment by yash-dani** [comment]
Hey @kulinseth @gpomeranz your patch gives me the following compile error:
```
error: patch failed: aten/src/ATen/native/mps/operations/BinaryOps.mm:330
error: aten/src/ATen/native/mps/operations/BinaryOps.mm: patch does not apply
error: patch failed: aten/src/ATen/native/native_functions.yaml:8802
error: aten/src/ATen/native/native_functions.yaml: patch does not apply
error: patch failed: test/test_mps.py:8143
error: test/test_mps.py: patch does not apply
```
What would be the best way to resolve? Thanks!

**Comment by kulinseth** [comment]
@yash-dani , seems like the patch and your branch has diverged since I had created that patch. 
Can you try manually just applying the changes I suggested in the patch?

```
--- a/aten/src/ATen/native/mps/operations/BinaryOps.mm
+++ b/aten/src/ATen/native/mps/operations/BinaryOps.mm
@@ -330,5 +330,29 @@ Tensor floor_divide_mps(const Tensor& self, const Tensor& other) {
   mps::binaryOpTensor(self, other, Scalar(1.0), output, "logaddexp2_out_mps", logaddexp2_op_block);
 }
 
+TORCH_IMPL_FUNC(remainder_out_mps) (const Tensor& self, const Tensor& other, const Tensor& output) {
+  // torch.remainder(a, b) == a - a.div(b, rounding_mode="floor") * b
+  mps::BinaryOpBlock remainder_op_block = ^BinaryOpFn(cachedGraph, primaryCastTensor, secondaryCastTensor) {
+    MPSGraph* mpsGraph = cachedGraph->graph();
+    // Rounding is a no-op for integral types, and also a reasonable workaround
+    // For MPSGraph bug on Apple Silicon, that throws `Function floorOp_i64 was not found in the library`
+    // See https://github.com/pytorch/pytorch/issues/84995
+    auto divTensor =  [mpsGraph divisionWithPrimaryTensor:primaryCastTensor
+                                          secondaryTensor:secondaryCastTensor
+                                                     name:nil];
+    //if (([divTensor dataType] & MPSDataTypeFloatBit) != 0) {
+      divTensor =  [mpsGraph floorWithTensor:divTensor name:nil];
+    //}
+    auto mulTensor = [mpsGraph multiplicationWithPrimaryTensor:divTensor
+                                               secondaryTensor:secondaryCastTensor
+                                                          name:nil];
+    return [mpsGraph subtractionWithPrimaryTensor:primaryCastTensor
+                                       secondaryTensor:mulTensor
+                                           name: nil];
+    };
+  mps::binaryOpTensor(self, other, Scalar(1.0), output, "remainder_out_mps", remainder_op_block);
+}
+
+
```

Basically it need to be added to the BinaryOps.mm

**Comment by kulinseth** [comment]
Thanks @gpomeranz for giving it a try? What issue did you run into ?

**Comment by gpomeranz** [comment]
@kulinseth I tried to manually apply the patches in the latest branch of pytorch, as I've never applied a patch to a github repo before. I unfortunately was not able to compile pytorch, so it stopped for me there. I don't remember what the error was but I believe it has something to do with my clang compiler (even though I use the latest brew one). I will try to redownload pytorch and apply the patch and will post here what my error was. 

I think if @yash-dani, who has more experience, manually adds the patch and tries to compile it will succeed.

**Comment by gpomeranz** [comment]
This is the error that I get when manually adding the patch and compiling pytorch:

/opt/homebrew/opt/llvm/bin/../include/c++/v1/__algorithm/iterator_operations.h:131:12: error: calling a private constructor of class 'c10::impl::ListElementReference<at::Tensor, std::__wrap_iter<c10::IValue *>>'
    return *std::forward<_Iter>(__i);

**Comment by jinsu35** [comment]
> @yash-dani , seems like the patch and your branch has diverged since I had created that patch. Can you try manually just applying the changes I suggested in the patch?
> 
> ```
> --- a/aten/src/ATen/native/mps/operations/BinaryOps.mm
> +++ b/aten/src/ATen/native/mps/operations/BinaryOps.mm
> @@ -330,5 +330,29 @@ Tensor floor_divide_mps(const Tensor& self, const Tensor& other) {
>    mps::binaryOpTensor(self, other, Scalar(1.0), output, "logaddexp2_out_mps", logaddexp2_op_block);
>  }
>  
> +TORCH_IMPL_FUNC(remainder_out_mps) (const Tensor& self, const Tensor& other, const Tensor& output) {
> +  // torch.remainder(a, b) == a - a.div(b, rounding_mode="floor") * b
> +  mps::BinaryOpBlock remainder_op_block = ^BinaryOpFn(cachedGraph, primaryCastTensor, secondaryCastTensor) {
> +    MPSGraph* mpsGraph = cachedGraph->graph();
> +    // Rounding is a no-op for integral types, and also a reasonable workaround
> +    // For MPSGraph bug on Apple Silicon, that throws `Function floorOp_i64 was not found in the library`
> +    // See https://github.com/pytorch/pytorch/issues/84995
> +    auto divTensor =  [mpsGraph divisionWithPrimaryTensor:primaryCastTensor
> +                                          secondaryTensor:secondaryCastTensor
> +                                                     name:nil];
> +    //if (([divTensor dataType] & MPSDataTypeFloatBit) != 0) {
> +      divTensor =  [mpsGraph floorWithTensor:divTensor name:nil];
> +    //}
> +    auto mulTensor = [mpsGraph multiplicationWithPrimaryTensor:divTensor
> +                                               secondaryTensor:secondaryCastTensor
> +                                                          name:nil];
> +    return [mpsGraph subtractionWithPrimaryTensor:primaryCastTensor
> +                                       secondaryTensor:mulTensor
> +                                           name: nil];
> +    };
> +  mps::binaryOpTensor(self, other, Scalar(1.0), output, "remainder_out_mps", remainder_op_block);
> +}
> +
> +
> ```
> 
> Basically it need to be added to the BinaryOps.mm

You do need this conditional block to handle a MPS bug. 
```
if (([divTensor dataType] & MPSDataTypeFloatBit) != 0) {
    divTensor =  [mpsGraph floorWithTensor:divTensor name:nil];
}
```
I applied the patch to the latest of master in PR #92139

**Comment by gpomeranz** [comment]
I can confirm that the above change indeed fixes the build problems and I was able to build pytorch from source using this code addition to the latest release. 

Looking forward to have this incorporated so that I can run my gpu accelerated code in R!

**Comment by vishank97** [comment]
Hi,
Any update on the fix? I'm still getting the error and I don't if there's a fix available yet. @yash-dani @gpomeranz @kulinseth

**Comment by kulinseth** [comment]
We have pulled in the remainder PR.

**Comment by kulinseth** [comment]
@yash-dani , please re-open if you think we missed something.

**Pull request closed without merging**